### DISCUSSION TITLE
Can I have a theme selector in revealjs menu?

### DISCUSSION THREAD

**Discussion by TLouf** [quarto-dev/quarto-cli#5831] [Q&A]
### Description

I have tried several things but could not utilize [reveal.js-menu](https://github.com/denehyg/reveal.js-menu)'s theme switching functionality. I have set:

```yml
menu:
      themes: true
```
in my config file, of course, but also tried including custom css in a folder, tweaking `themesPath` so that the plugin could reach it. But in the end, I couldn't manage to have the `Themes` tab appear in the menu.

Would anyone have a solution to propose? The way may be a bit hacky but I would say that there should be one, right?

Thanks for the help! :)

**Comment by cderv** [comment]
I don't think we allow for configuring the built in menu plugin with YAML option. 

A current workaround would be to set `menu: false` to not include the default menu, and then add another tweaked menu plugin packaged as an extension (https://quarto.org/docs/extensions/revealjs.html) but with other default. 

This would be a Feature request to allow changing some settings for the built in plugins

**Reply by TLouf** [reply]
The thing is, setting `transitions: true` does have the expected effect, for instance. So I have the feeling that I can somehow make the theme selector work, albeit with some hack to make it find the CSS files.

But thanks, disabling the already-packaged one to use one I include myself surely works too, although not ideal.

**Comment by mcanouil** [comment]
In theory, you can, but it's likely to mess up all Quarto (S)CSS and you need to have the (S)CSS locally.

```
---
title: "Presentation"
format:
 revealjs:
   theme: dark
   menu: {
     themes: [
       {
         name: 'Black',
         theme: 'dist/theme/beige.css'
       }
     ]
   }
---
```

See <https://github.com/denehyg/reveal.js-menu>.

**Reply by cderv** [reply]
Oh great ! I couldn't make it works. I tried to transform the JSON structure to YAML I guess. This is great if this works ! 

I looked at the code and it seemed it wasn't possible to configure the built in plugins like menu because we use it and tweak it already. It is possible I missed the code part that does merge the config. 

Thanks for sharing @mcanouil !

**Reply by mcanouil** [reply]
To be fair, @ArthurData brought this to my attention few months back 😅

**Reply by cderv** [reply]
And passing JSON content as a value in a YAML field works without issue ? No need for the multline syntax, 
```
menu: |
	{ ...
    }
```

or any raw attributes ? 

Maybe not because Quarto itself is handling the YAML and not pandoc 🤔

**Reply by mcanouil** [reply]
I copied the code and did not spend too much time on this so far, so I don't know if it's working or not. 🤷‍♂️

**Reply by cderv** [reply]
Oh ok. Maybe this should be tested then before considering it an answer. I know some of the configuration for revealjs can't be passed using YAML unfortunately. 

And also, maybe my code reading is right in my first answer above: https://github.com/quarto-dev/quarto-cli/discussions/5831#discussioncomment-6109684 

I am really not sure we allow this unfortunately, and I thought you were sharing that it is working. I am a bit lost now... 😅

**Reply by ArthurData** [reply]
Hello,

The solution "works" but not quite as planned, I think.

![Jun-12-2023 20-31-36](https://github.com/quarto-dev/quarto-cli/assets/35060481/0eeaea43-1be0-4fce-993f-bc308f9df1ae)


However, I get an error when I pass the JSON as a value :

```
---
title: "Presentation"
format:
  revealjs: 
    theme: dark
    menu: |
      {
        themes: [
          { 
            name: 'Red',
            theme: 'dist/theme/red.css'
          }
        ]
      }
---

# ...
```

```
ERROR: Render failed due to invalid YAML.
```

**Reply by cderv** [reply]
Thanks for testing ! 

As @mcanouil said, using a theme file as `css` with this extension will probably break things as other configuration for Quarto are not merged probably. When the red.css is applied, it must be replacing some global rules. 🤔

**Reply by mcanouil** [reply]
@cderv @ArthurData is the one who told be about breaking things and showing me the possibility of theme menu in "raw" Reveal.js 😜

**Reply by cderv** [reply]
😆 ok ! I am still puzzled by this raw content pass to YAML field.... 🤔  I believe this is working as a side effect, and not exactly as we think it is. 

YAML has a Flow style availble, including Flow mapping, which allow to use `{ } ` instead of indentation. Search FLow Styles in spec: https://yaml.org/spec/1.2.2
> YAML also has [flow styles](https://yaml.org/spec/1.2.2/#flow-style-productions), using explicit [indicators](https://yaml.org/spec/1.2.2/#indicator-characters) rather than [indentation](https://yaml.org/spec/1.2.2/#indentation-spaces) to denote scope. The [flow sequence](https://yaml.org/spec/1.2.2/#flow-sequences) is written as a [comma](https://yaml.org/spec/1.2.2/#flow-collection-styles) separated list within [square](https://yaml.org/spec/1.2.2/#flow-sequences) [brackets](https://yaml.org/spec/1.2.2/#flow-sequences). In a similar manner, the [flow mapping](https://yaml.org/spec/1.2.2/#flow-mappings) uses [curly](https://yaml.org/spec/1.2.2/#flow-mappings) [braces](https://yaml.org/spec/1.2.2/#flow-mappings).

This means that 
````yaml
format:
 revealjs:
   theme: dark
   menu: {
     themes: [
       {
         name: 'Black',
         theme: 'dist/theme/beige.css'
       }
     ]
   }
````

is the same as 
````yaml
format:
 revealjs:
   theme: dark
   menu:
     themes:
      - name: 'Black'
         theme: 'dist/theme/beige.css'
````

It seems we are correctly merging the config by formatting internally to the right JSON for Revealjs conf. 

This solves for me the puzzle, and why using `menu: |` does not work, because passing Raw JSON is not permitted. 

Thanks for the question and answer, as I learnt about YAML flow styles

**Reply by TLouf** [reply]
Thanks a lot for the investigations everyone! @ArthurData I can't make your solution work, I had tried something similar already, but can't figure out where the `dist` folder should be located in my working directory, as I don't know how quarto resolves the path `dist/theme/red.css`. I basically copied the CSS all over the place, but to no avail, there must be something dumb I'm missing.
